oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/TraverseAndSortTask.java

```diff
@@ -76,57 +84,82 @@
+                        BlobStore blobStore, File storeDir, boolean compressionEnabled,
+                        Queue<String> completedTasks, Queue<Callable<List<File>>> newTasksQueue,
+                        Phaser phaser, NodeStateEntryTraverserFactory nodeStateEntryTraverserFactory) {
+        this.taskID = "TWS-" + nodeStates.getId();
         this.nodeStates = nodeStates;
-        this.entryWriter = entryWriter;
+        this.lastModifiedLowerBound = nodeStates.getDocumentModificationRange().getLastModifiedLowerBound();
+        this.lastModifiedUpperBound = nodeStates.getDocumentModificationRange().getLastModifiedUpperBound();
+        this.blobStore = blobStore;
+        this.entryWriter = new NodeStateEntryWriter(blobStore);
         this.storeDir = storeDir;
         this.compressionEnabled = compressionEnabled;
-        this.comparator = (e1, e2) -> pathComparator.compare(e1.getPathElements(), e2.getPathElements());
+        this.comparator = comparator;
+        this.completedTasks = completedTasks;
+        this.newTasksQueue = newTasksQueue;
+        this.phaser = phaser;
+        this.nodeStateEntryTraverserFactory = nodeStateEntryTraverserFactory;
+        phaser.register();
+        log.debug("Task {} registered to phaser", taskID);
     }
 
-    @Override
-    public File createSortedStoreFile() throws IOException {
-        logFlags();
-        configureMemoryListener();
-        sortWorkDir = createdSortWorkDir(storeDir);
-        writeToSortedFiles();
-        return sortStoreFile();
+    public List<File> call() {
+        try {
+            logFlags();
+            configureMemoryListener();
+            sortWorkDir = MultithreadedTraversalWithSortStrategy.DirectoryHelper.createdSortWorkDir(storeDir, taskID,
+                    lastModifiedLowerBound);
+            writeToSortedFiles();
+            log.info("Completed task {}", taskID);
+            phaser.arriveAndDeregister();
+            completedTasks.add(taskID);
+            return sortedFiles;
+        } catch (IOException e) {
+            log.error(taskID + " could not complete download ", e);
+        }
+        return Collections.emptyList();
     }
 
-    @Override
     public long getEntryCount() {
         return entryCount;
     }
 
-    private File sortStoreFile() throws IOException {
-        log.info("Proceeding to perform merge of {} sorted files", sortedFiles.size());
-        Stopwatch w = Stopwatch.createStarted();
-        File sortedFile = new File(storeDir, getSortedStoreFileName(compressionEnabled));
-        try(BufferedWriter writer = createWriter(sortedFile, compressionEnabled)) {
-            Function<String, NodeStateHolder> func1 = (line) -> line == null ? null : new SimpleNodeStateHolder(line);
-            Function<NodeStateHolder, String> func2 = holder -> holder == null ? null : holder.getLine();
-            ExternalSort.mergeSortedFiles(sortedFiles,
```

Comment:
I don't currently see where we sort... we do need to sort, right?

Reply:
Yes, individual files are sorted in the method - `sortAndSaveBatch` in this class itself.
The sorted merge is done in `MultithreadedTraversalWithSortStrategy#sortStoreFile`